main/acle.md

```diff
@@ -7499,6 +7653,1415 @@ This is because the ACLE intrinsic calls do not imply a particular
+defines various intrinsics. The support functions have
+[external linkage](#external-linkage), like standard C functions such
+as `memcpy` do. However, as noted in [Intrinsics](#intrinsics), it is
+unspecified whether the intrinsics are functions and, if so, what
+linkage they have.
+
+There are many more intrinsics than support functions, so everything
+described in this section is an intrinsic unless otherwise noted.
+
+If an intrinsic is implemented as a macro rather than a function,
+the macro must behave “as if” it had the prototype and attributes
+specified in this section.
+
+### SME PSTATE functions
+
+#### Prototypes
+
+``` c
+  __attribute__((arm_streaming_compatible))
+  bool __arm_in_streaming_mode(void);
+
+  // Function with external linkage.
+  __attribute__((arm_streaming_compatible))
+  void __arm_za_disable(void);
+```
+
+#### Semantics
+
+<!-- Do not remove the following `span`, they are needed to create an
+anchor that can be referred via an internal hyperlink to the paragraph
+following it. --><span id="__arm_in_streaming_mode"></span>
+
+**`__arm_in_streaming_mode()`**
+
+> This call returns true if the [abstract machine](#abstract-machine) is
+> in streaming mode. It always returns true when called from [streaming
+> statements](#streaming-statement) and it always return false when
+> called from [non-streaming statements](#non-streaming-statement).
+> However, the call is not semantically a constant expression even in
+> those cases.
+
+<!-- Do not remove the following `span`, they are needed to create an
+anchor that can be referred via an internal hyperlink to the paragraph
+following it. --><span id="__arm_za_disable"></span>
+
+**`__arm_za_disable()`**
+
+> This call commits any pending lazy save and turns ZA off;
+> see [[AAPCS64]](#AAPCS64) for details.
+>
+> Note: since this is a [private-ZA](#private-za) function,
+> it is valid (though pointless) to call it from a function that has
+> [ZA state](#za-state). The compiler would simply reenable ZA after the
+> call and reload the saved ZA state.
+
+### SME ZA state assertions
+
+#### Prototypes
+
+``` c
+  __attribute__((arm_streaming_compatible, arm_shared_za))
+  void svundef_za();
+```
+
+#### Semantics
+
+**`svundef_za()`**
+
+> This call asserts that ZA does not currently have any useful data.
+> Semantically, it fills ZA with unpredictable data, although from
+> a quality of implementation perspective, it should not generate
+> any object code.
+>
+> The call can be used as an optimization hint to the compiler, so that
+> the compiler does not insert unnecessary code to save and restore the
+> current ZA contents. The call might also be useful for static analysis.
+
+### SME instruction intrinsics
+
+#### Common rules
+
+The intrinsics in this section have the following properties in common:
+
+*   Every argument named `tile`, `slice_offset` or `tile_mask` must
+    be an integer constant expression.
```

Comment:
I just came across this ACLE draft. What was the rationale for making it the user's responsibility to index into ZA storage explicitly, instead of creating sizeless types analogous to `svint32_t`, `svfloat64_t`, etc.? We have done [some work](https://github.com/Huawei-PTLab/llvm-project/tree/scalable-matrix) in that direction, and extended ACLE to provide types such as `smint32_t` and `smfloat64_t`. Did you consider that approach at all?

Reply:
Hi  Thanks for the pointer, this looks like really nice work.  Yeah, we did consider using explicit C/C++ objects to represent the arrays, but in the end, we thought it would be better for the low-level ACLE to maintain a more direct mapping to the instructions.  There were several reasons for this:

* It seemed likely that programmers writing code specifically for SME would be aware of the number of available tiles for a given element type, would know how the tiles are arranged, and would want to use that knowledge to hide accumulation latencies.  We therefore wanted to let the programmer specify tile numbers directly if they wanted to.  In other words, we thought SME programmers would be thinking of ZA as a whole (unlike SVE programmers, who would think of vectors as individual objects rather than as 1 out of 32 vector registers).
* Although compilers are improving all the time, there are still cases where they make bad RA decisions for vector code, especially when inputs and outputs need to be tied.  The danger with treating matrices like vectors is that the cost of any compiler mistakes would be O(VL) rather than O(1).  And it can be quite difficult to work around these issues in the source code.  This might be a particular issue when “reinterpreting” 8 64-bit tiles as 1 8-bit tile (for example), since the 8 tiles would need to be in a particular order.
* If the tiles were normal C/C++ objects, a function wouldn't be able to return a scalable matrix object (in ZA) at the same time as returning a normal object (in GPRs or FPRs) unless we provided some way of putting scalable objects into structures/classes.  That would be a good thing long term, but it seems difficult to do, since C++ has a fundamental assumption that sizeof is a constant expression.
* Complex operations might be split over several subroutines.  We thought that, in those cases, the routines would be sharing ZA as a whole, so having a single `arm_shared_za` attribute seemed more convenient than having to pass multiple tiles around.

However, the above is all about the low-level ACLE interface.  I don't think this has to be an “either/or”.  It would be useful to have higher-level features too, where the compiler does more of the work.  Also, having scalable matrix types in LLVM IR sounds like it should be a good fit for Florian's Clang matrix extensions.

Reply:
Hi @rsandifo-arm, thank you for the thoughtful answer. I agree with you that this doesn't have to be an "either/or"; perhaps it's even possible to implement this ACLE draft on top of our work. You have raised very good points about register allocation decision and reinterpretation; we found that we would need to teach LLVM's register allocator about sub-tiles in SME to make it do a better job, and if we were to support a reinterpretation intrinsic, it would put even more constraints on RA.

About a function's inability to return a scalable matrix object at the same time as returning a normal object, wouldn't the same limitation apply to scalable vector objects as well? Or has that problem already been solved?

Reply:
Hi @bryanpkc : No, you're right, the same restriction applies to scalable vector objects as well.  The `x2_t`, `x3_t` and `x4_t` types make it possible to return up to 4 vectors at a time (although all vectors need to be the same time, which would introduce some awkward reinterprets if someone wanted to return, say, a vector of `float`s and a vector of `uint32_t`s).  But there is currently no way of returning both a scalable vector and a scalar by value, or even a scalable vector and a scalable predicate.  It's an unfortunate restriction.

It is of course possible to return things by reference, but it would be better if that wasn't necessary.

Reply:
Hi both.

@bryanpkc  - thank you for raising your concerns. Please correct if I am wrong, but my understanding is that you are happy with explanations that @rsandifo-arm provided in their answer. If that is the case, are you happy for us to proceed with merging the specs for the SME ACLE as they are in this patch? We will be open to further improvements, if needed.

@rsandifo-arm  - your answer is quite useful as it provides a justification of the current shape of the SME ACLE. Do you mind adding it to and SME-specific design document in the `design_documents` folder?

Kind regards,

Francesco

Reply:
@fpetrogalli Sorry for the late response. I am happy with the explanations from @rsandifo-arm and have no more comment on this point.

Reply:
@bryanpkc - thank you for the comments. Are you happy for me to add you as a contributor to the [list of all-contributors](https://github.com/ARM-software/acle/#contributors-) for your review?